pool.go

```diff
@@ -24,28 +24,85 @@ import (
 )
 
 type modulePool[T interfaces.Signature] struct {
-	pool sync.Pool
-	new  func() (*module[T], error)
+	pool    sync.Pool
+	maxSize uint32
+	new     func(*module[T]) (*module[T], error)
+	close   func(*module[T])
+	ch      chan *module[T] // for buffered channel impl
 }
 
-func newModulePool[T interfaces.Signature](ctx context.Context, template *template[T]) *modulePool[T] {
-	return &modulePool[T]{
-		new: func() (*module[T], error) {
-			return newModule[T](ctx, template)
-		},
+func newModulePool[T interfaces.Signature](ctx context.Context, template *template[T], maxSize uint32) *modulePool[T] {
+	if maxSize == 0 {
+		// If size = 0, use a standard sync.Pool + golang's runtime finalizers
+		// to make sure the close function gets called eventually.
+		return &modulePool[T]{
+			maxSize: maxSize,
+			new: func(m *module[T]) (*module[T], error) {
+				m.SetFinalizer()
+				return newModule[T](ctx, template)
```

Comment:
The `ctx` being used here is the context provided in calls to `newModulePool`. I'm not familiar with where the context passed to `newModulePool` ultimately comes from, but if it comes from the user, the context might be dead by the time this pool wants to allocate new modules in the future.